// @okexchain/javascript-sdk/lib/client.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.OKEXChainClient = exports.GetClient = void 0;

var crypto = _interopRequireWildcard(require("./crypto"));

var _transaction = _interopRequireDefault(require("./transaction"));

var _httpProxy = _interopRequireDefault(require("./httpProxy"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj["default"] = obj; return newObj; } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

var apiPath = {
  //rest server
  txs: "/okexchain/v1/txs",
  queryAccount: "/okexchain/v1/auth/accounts"
};
var defaultChainId = "okexchaintestnet-1"; //"test-chain-2yOdpj"

var bech32Head = "okexchain";
var mode = "block";
var nativeDenom = "okt"; // t

var defaultFee = {
  amount: [{
    amount: "0.020000000000000000",
    denom: nativeDenom
  }],
  gas: "200000"
};
var precision = 18;

var GetClient =
/*#__PURE__*/
function () {
  var _ref = _asyncToGenerator(
  /*#__PURE__*/
  regeneratorRuntime.mark(function _callee(privateKey, url) {
    var client;
    return regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            client = new OKEXChainClient(url);
            client.setAccountInfo(privateKey);
            return _context.abrupt("return", client);

          case 3:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));

  return function GetClient(_x, _x2) {
    return _ref.apply(this, arguments);
  };
}();
/**
 * The OKEXChain client.
 */


exports.GetClient = GetClient;

var OKEXChainClient =
/*#__PURE__*/
function () {
  /**
   * @param {string} url
   * @param {string} chainId
   */
  function OKEXChainClient(url, chainId) {
    _classCallCheck(this, OKEXChainClient);

    if (!url) {
      throw new Error("null url");
    }

    this.httpClient = new _httpProxy["default"](url);
    this.mode = mode;
    this.chainId = chainId || defaultChainId;
  }
  /**
   * set the mode when send transaction
   * @param {string} mode block|sync|async
   */


  _createClass(OKEXChainClient, [{
    key: "setMode",
    value: function () {
      var _setMode = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee2(m) {
        return regeneratorRuntime.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                this.mode = m;

              case 1:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function setMode(_x3) {
        return _setMode.apply(this, arguments);
      }

      return setMode;
    }()
    /**
     * set the chainId when send transaction
     * @param {string} id
     */

  }, {
    key: "setChainId",
    value: function () {
      var _setChainId = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee3(id) {
        return regeneratorRuntime.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                this.chainId = id;

              case 1:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function setChainId(_x4) {
        return _setChainId.apply(this, arguments);
      }

      return setChainId;
    }()
    /**
     * @param {string} privateKey
     * @return {OKEXChainClient}
     */

  }, {
    key: "setAccountInfo",
    value: function () {
      var _setAccountInfo = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee4(privateKey) {
        var address, data;
        return regeneratorRuntime.wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                if (!(privateKey !== this.privateKey)) {
                  _context4.next = 14;
                  break;
                }

                address = crypto.getAddressFromPrivateKey(privateKey, bech32Head);

                if (address) {
                  _context4.next = 4;
                  break;
                }

                throw new Error("invalid privateKey: " + privateKey);

              case 4:
                if (!(address === this.address)) {
                  _context4.next = 6;
                  break;
                }

                return _context4.abrupt("return", this);

              case 6:
                this.privateKey = privateKey;
                this.address = address;
                _context4.next = 10;
                return this.getAccount(address);

              case 10:
                data = _context4.sent;
                _context4.next = 13;
                return this.getAccountNumberFromAccountInfo(data);

              case 13:
                this.account_number = _context4.sent;

              case 14:
                return _context4.abrupt("return", this);

              case 15:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function setAccountInfo(_x5) {
        return _setAccountInfo.apply(this, arguments);
      }

      return setAccountInfo;
    }()
    /**
     * Send SendTransaction.
     * @param {String} to To Address
     * @param {Number} amount Coin Quantity
     * @param {String} denom Coin Name
     * @param {String} memo
     * @param {Number} sequenceNumber
     * @return {Object} response
     */

  }, {
    key: "sendSendTransaction",
    value: function () {
      var _sendSendTransaction = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee5(to, amount, denom) {
        var memo,
            sequenceNumber,
            coin,
            msg,
            signMsg,
            signedTx,
            res,
            _args5 = arguments;
        return regeneratorRuntime.wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                memo = _args5.length > 3 && _args5[3] !== undefined ? _args5[3] : "";
                sequenceNumber = _args5.length > 4 && _args5[4] !== undefined ? _args5[4] : null;
                coin = {
                  amount: this.formatNumber(amount),
                  denom: denom
                };
                msg = [{
                  type: "okexchain/token/MsgTransfer",
                  value: {
                    amount: [coin],
                    from_address: this.address,
                    to_address: to
                  }
                }];
                signMsg = msg;
                _context5.next = 7;
                return this.buildTransaction(msg, signMsg, memo, defaultFee, sequenceNumber);

              case 7:
                signedTx = _context5.sent;
                _context5.next = 10;
                return this.sendTransaction(signedTx);

              case 10:
                res = _context5.sent;
                return _context5.abrupt("return", res);

              case 12:
              case "end":
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));

      function sendSendTransaction(_x6, _x7, _x8) {
        return _sendSendTransaction.apply(this, arguments);
      }

      return sendSendTransaction;
    }()
    /**
     * Send CancelOrderTransaction.
     * @param {String} orderId
     * @param {String} memo
     * @param {Number} sequenceNumber
     * @return {Object} response
     */

  }, {
    key: "sendCancelOrderTransaction",
    value: function () {
      var _sendCancelOrderTransaction = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee6(orderId) {
        var memo,
            sequenceNumber,
            orderIdList,
            _args6 = arguments;
        return regeneratorRuntime.wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                memo = _args6.length > 1 && _args6[1] !== undefined ? _args6[1] : "";
                sequenceNumber = _args6.length > 2 && _args6[2] !== undefined ? _args6[2] : null;
                orderIdList = [orderId];
                return _context6.abrupt("return", this.sendCancelOrdersTransaction(orderIdList, memo, sequenceNumber));

              case 4:
              case "end":
                return _context6.stop();
            }
          }
        }, _callee6, this);
      }));

      function sendCancelOrderTransaction(_x9) {
        return _sendCancelOrderTransaction.apply(this, arguments);
      }

      return sendCancelOrderTransaction;
    }()
  }, {
    key: "sendCancelOrdersTransaction",
    value: function () {
      var _sendCancelOrdersTransaction = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee7(orderIdList) {
        var memo,
            sequenceNumber,
            msg,
            signMsg,
            signedTx,
            res,
            _args7 = arguments;
        return regeneratorRuntime.wrap(function _callee7$(_context7) {
          while (1) {
            switch (_context7.prev = _context7.next) {
              case 0:
                memo = _args7.length > 1 && _args7[1] !== undefined ? _args7[1] : "";
                sequenceNumber = _args7.length > 2 && _args7[2] !== undefined ? _args7[2] : null;
                msg = [];
                signMsg = [];
                msg.push({
                  type: "okexchain/order/MsgCancel",
                  value: {
                    order_ids: orderIdList,
                    sender: this.address
                  }
                });
                signMsg = msg;
                _context7.next = 8;
                return this.buildTransaction(msg, signMsg, memo, defaultFee, sequenceNumber);

              case 8:
                signedTx = _context7.sent;
                _context7.next = 11;
                return this.sendTransaction(signedTx);

              case 11:
                res = _context7.sent;
                return _context7.abrupt("return", res);

              case 13:
              case "end":
                return _context7.stop();
            }
          }
        }, _callee7, this);
      }));

      function sendCancelOrdersTransaction(_x10) {
        return _sendCancelOrdersTransaction.apply(this, arguments);
      }

      return sendCancelOrdersTransaction;
    }()
    /*
     * format number
     */

  }, {
    key: "formatNumber",
    value: function formatNumber(num) {
      var str = String(num);
      var retStr = '';

      if (str.indexOf('.') >= 0) {
        if (str.split('.')[1].length > precision) {
          throw new Error("The actual received decimal precision is " + str.split('.')[1].length + ", and the expected is " + precision);
        } else if (str.split('.')[1].length == precision) {
          retStr = str;
        } else {
          var appendix = '';
          var len = precision - str.split('.')[1].length;

          for (var i = 0; i < len; i++) {
            appendix += '0';
          }

          retStr = str + appendix;
        }
      } else {
        retStr += "".concat(str, ".000000000000000000");
      }

      return retStr;
    }
  }, {
    key: "sendPlaceOrderTransaction",

    /**
     * Send PlaceOrderTransaction.
     * @param {String} product
     * @param {String} side
     * @param {Number} price
     * @param {Number} quantity
     * @param {Number} memo
     * @param {Number} sequence
     * @return {Object} response
     */
    value: function () {
      var _sendPlaceOrderTransaction = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee8(product, side, price, quantity) {
        var memo,
            sequence,
            order_items,
            _args8 = arguments;
        return regeneratorRuntime.wrap(function _callee8$(_context8) {
          while (1) {
            switch (_context8.prev = _context8.next) {
              case 0:
                memo = _args8.length > 4 && _args8[4] !== undefined ? _args8[4] : "";
                sequence = _args8.length > 5 && _args8[5] !== undefined ? _args8[5] : null;
                order_items = [{
                  price: this.formatNumber(price),
                  product: product,
                  quantity: this.formatNumber(quantity),
                  side: side
                }];
                return _context8.abrupt("return", this.sendPlaceOrdersTransaction(order_items, memo, sequence));

              case 4:
              case "end":
                return _context8.stop();
            }
          }
        }, _callee8, this);
      }));

      function sendPlaceOrderTransaction(_x11, _x12, _x13, _x14) {
        return _sendPlaceOrderTransaction.apply(this, arguments);
      }

      return sendPlaceOrderTransaction;
    }()
  }, {
    key: "sendPlaceOrdersTransaction",
    value: function () {
      var _sendPlaceOrdersTransaction = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee9(order_items) {
        var memo,
            sequence,
            placeOrderMsg,
            signMsg,
            signedTx,
            res,
            _args9 = arguments;
        return regeneratorRuntime.wrap(function _callee9$(_context9) {
          while (1) {
            switch (_context9.prev = _context9.next) {
              case 0:
                memo = _args9.length > 1 && _args9[1] !== undefined ? _args9[1] : "";
                sequence = _args9.length > 2 && _args9[2] !== undefined ? _args9[2] : null;
                placeOrderMsg = [{
                  type: "okexchain/order/MsgNew",
                  value: {
                    order_items: order_items,
                    sender: this.address
                  }
                }];
                signMsg = placeOrderMsg;
                _context9.next = 6;
                return this.buildTransaction(placeOrderMsg, signMsg, memo, defaultFee, sequence);

              case 6:
                signedTx = _context9.sent;
                _context9.next = 9;
                return this.sendTransaction(signedTx);

              case 9:
                res = _context9.sent;
                return _context9.abrupt("return", res);

              case 11:
              case "end":
                return _context9.stop();
            }
          }
        }, _callee9, this);
      }));

      function sendPlaceOrdersTransaction(_x15) {
        return _sendPlaceOrdersTransaction.apply(this, arguments);
      }

      return sendPlaceOrdersTransaction;
    }()
    /**
     * Build Transaction for sending to okexchain.
     * @param {Object} msg
     * @param {Object} signMsg
     * @param {String} memo
     * @param {String} fee
     * @param {Number} sequenceNumber
     * @return {Transaction} Transaction object
     */

  }, {
    key: "buildTransaction",
    value: function () {
      var _buildTransaction = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee10(msg, signMsg) {
        var memo,
            fee,
            sequenceNumber,
            accountInfo,
            params,
            tx,
            _args10 = arguments;
        return regeneratorRuntime.wrap(function _callee10$(_context10) {
          while (1) {
            switch (_context10.prev = _context10.next) {
              case 0:
                memo = _args10.length > 2 && _args10[2] !== undefined ? _args10[2] : "";
                fee = _args10.length > 3 && _args10[3] !== undefined ? _args10[3] : null;
                sequenceNumber = _args10.length > 4 && _args10[4] !== undefined ? _args10[4] : null;

                if (!(!this.account_number || !sequenceNumber)) {
                  _context10.next = 13;
                  break;
                }

                _context10.next = 6;
                return this.getAccount();

              case 6:
                accountInfo = _context10.sent;
                _context10.next = 9;
                return this.getSequenceNumberFromAccountInfo(accountInfo);

              case 9:
                sequenceNumber = _context10.sent;
                _context10.next = 12;
                return this.getAccountNumberFromAccountInfo(accountInfo);

              case 12:
                this.account_number = _context10.sent;

              case 13:
                params = {
                  account_number: parseInt(this.account_number),
                  chain_id: this.chainId,
                  memo: memo,
                  msg: msg,
                  sequence: sequenceNumber,
                  fee: fee
                };
                tx = new _transaction["default"](params);
                return _context10.abrupt("return", tx.sign(this.privateKey, signMsg));

              case 16:
              case "end":
                return _context10.stop();
            }
          }
        }, _callee10, this);
      }));

      function buildTransaction(_x16, _x17) {
        return _buildTransaction.apply(this, arguments);
      }

      return buildTransaction;
    }()
    /**
     * send transaction to OKEXChain.
     * @param {signedTx} tx signed Transaction object
     * @param {Boolean} mode use synchronous mode, optional
     * @return {Object} response (success or fail)
     */

  }, {
    key: "sendTransaction",
    value: function () {
      var _sendTransaction = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee11(signedTx) {
        var buf, opts;
        return regeneratorRuntime.wrap(function _callee11$(_context11) {
          while (1) {
            switch (_context11.prev = _context11.next) {
              case 0:
                buf = signedTx.serializeTransactionWithJson(this.mode);
                console.log(buf);
                opts = {
                  data: buf,
                  headers: {
                    "content-type": "text/plain"
                  }
                };
                return _context11.abrupt("return", this.httpClient.send("post", "".concat(apiPath.txs), null, opts));

              case 4:
              case "end":
                return _context11.stop();
            }
          }
        }, _callee11, this);
      }));

      function sendTransaction(_x18) {
        return _sendTransaction.apply(this, arguments);
      }

      return sendTransaction;
    }()
    /**
     * get account
     * @param {String} address
     * @return {Object} result
     */

  }, {
    key: "getAccount",
    value: function () {
      var _getAccount = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee12() {
        var address,
            data,
            _args12 = arguments;
        return regeneratorRuntime.wrap(function _callee12$(_context12) {
          while (1) {
            switch (_context12.prev = _context12.next) {
              case 0:
                address = _args12.length > 0 && _args12[0] !== undefined ? _args12[0] : this.address;

                if (address) {
                  _context12.next = 3;
                  break;
                }

                throw new Error("address should not be falsy");

              case 3:
                _context12.prev = 3;
                _context12.next = 6;
                return this.httpClient.send("get", "".concat(apiPath.queryAccount, "/").concat(address));

              case 6:
                data = _context12.sent;
                return _context12.abrupt("return", data);

              case 10:
                _context12.prev = 10;
                _context12.t0 = _context12["catch"](3);
                return _context12.abrupt("return", null);

              case 13:
              case "end":
                return _context12.stop();
            }
          }
        }, _callee12, this, [[3, 10]]);
      }));

      function getAccount() {
        return _getAccount.apply(this, arguments);
      }

      return getAccount;
    }()
    /**
     * get balances from OKEXChain
     * @param {String} address
     * @return {Object} result
     */

  }, {
    key: "getBalance",
    value: function () {
      var _getBalance = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee13() {
        var address,
            data,
            _args13 = arguments;
        return regeneratorRuntime.wrap(function _callee13$(_context13) {
          while (1) {
            switch (_context13.prev = _context13.next) {
              case 0:
                address = _args13.length > 0 && _args13[0] !== undefined ? _args13[0] : this.address;
                _context13.prev = 1;
                _context13.next = 4;
                return this.getAccount(address);

              case 4:
                data = _context13.sent;
                return _context13.abrupt("return", this.getBalanceFromAccountInfo(data));

              case 8:
                _context13.prev = 8;
                _context13.t0 = _context13["catch"](1);
                return _context13.abrupt("return", []);

              case 11:
              case "end":
                return _context13.stop();
            }
          }
        }, _callee13, this, [[1, 8]]);
      }));

      function getBalance() {
        return _getBalance.apply(this, arguments);
      }

      return getBalance;
    }()
    /**
     * get balances from accountInfo Object
     * @param {Object} accountInfo optional address
     * @return {Object} result
     */

  }, {
    key: "getBalanceFromAccountInfo",
    value: function () {
      var _getBalanceFromAccountInfo = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee14(accountInfo) {
        return regeneratorRuntime.wrap(function _callee14$(_context14) {
          while (1) {
            switch (_context14.prev = _context14.next) {
              case 0:
                return _context14.abrupt("return", accountInfo.result.value.coins);

              case 1:
              case "end":
                return _context14.stop();
            }
          }
        }, _callee14);
      }));

      function getBalanceFromAccountInfo(_x19) {
        return _getBalanceFromAccountInfo.apply(this, arguments);
      }

      return getBalanceFromAccountInfo;
    }()
    /**
     * get SequenceNumber from OKEXChain
     * @param {String} address
     * @return {Number} sequenceNumber
     */

  }, {
    key: "getSequenceNumber",
    value: function () {
      var _getSequenceNumber = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee15() {
        var address,
            data,
            _args15 = arguments;
        return regeneratorRuntime.wrap(function _callee15$(_context15) {
          while (1) {
            switch (_context15.prev = _context15.next) {
              case 0:
                address = _args15.length > 0 && _args15[0] !== undefined ? _args15[0] : this.address;
                _context15.prev = 1;
                _context15.next = 4;
                return this.getAccount(address);

              case 4:
                data = _context15.sent;
                return _context15.abrupt("return", this.getSequenceNumberFromAccountInfo(data));

              case 8:
                _context15.prev = 8;
                _context15.t0 = _context15["catch"](1);
                return _context15.abrupt("return", null);

              case 11:
              case "end":
                return _context15.stop();
            }
          }
        }, _callee15, this, [[1, 8]]);
      }));

      function getSequenceNumber() {
        return _getSequenceNumber.apply(this, arguments);
      }

      return getSequenceNumber;
    }()
    /**
     * get SequenceNumber from accountInfo Object
     * @param {String} accountInfo
     * @return {Number} sequenceNumber
     */

  }, {
    key: "getSequenceNumberFromAccountInfo",
    value: function () {
      var _getSequenceNumberFromAccountInfo = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee16(accountInfo) {
        return regeneratorRuntime.wrap(function _callee16$(_context16) {
          while (1) {
            switch (_context16.prev = _context16.next) {
              case 0:
                return _context16.abrupt("return", accountInfo.result.value.sequence);

              case 1:
              case "end":
                return _context16.stop();
            }
          }
        }, _callee16);
      }));

      function getSequenceNumberFromAccountInfo(_x20) {
        return _getSequenceNumberFromAccountInfo.apply(this, arguments);
      }

      return getSequenceNumberFromAccountInfo;
    }()
    /**
     * get accountNumber from accountInfo Object
     * @param {String} accountInfo
     * @return {Number} accountNumber
     */

  }, {
    key: "getAccountNumberFromAccountInfo",
    value: function () {
      var _getAccountNumberFromAccountInfo = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee17(accountInfo) {
        return regeneratorRuntime.wrap(function _callee17$(_context17) {
          while (1) {
            switch (_context17.prev = _context17.next) {
              case 0:
                return _context17.abrupt("return", accountInfo.result.value.account_number);

              case 1:
              case "end":
                return _context17.stop();
            }
          }
        }, _callee17);
      }));

      function getAccountNumberFromAccountInfo(_x21) {
        return _getAccountNumberFromAccountInfo.apply(this, arguments);
      }

      return getAccountNumberFromAccountInfo;
    }()
    /**
     * Send TokenIssueTransaction.
     * @param {String} symbol
     * @param {String} whole_name
     * @param {String} total_supply
     * @param {Boolean} mintable
     * @param {String} description
     * @param {String} memo
     * @param {Number} sequenceNumber
     * @return {Object} response
     */

  }, {
    key: "sendTokenIssueTransaction",
    value: function () {
      var _sendTokenIssueTransaction = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee18(symbol, whole_name, total_supply) {
        var mintable,
            description,
            memo,
            sequenceNumber,
            msg,
            signedTx,
            res,
            _args18 = arguments;
        return regeneratorRuntime.wrap(function _callee18$(_context18) {
          while (1) {
            switch (_context18.prev = _context18.next) {
              case 0:
                mintable = _args18.length > 3 && _args18[3] !== undefined ? _args18[3] : false;
                description = _args18.length > 4 && _args18[4] !== undefined ? _args18[4] : '';
                memo = _args18.length > 5 && _args18[5] !== undefined ? _args18[5] : '';
                sequenceNumber = _args18.length > 6 && _args18[6] !== undefined ? _args18[6] : null;
                msg = [{
                  type: "okexchain/token/MsgIssue",
                  value: {
                    description: description,
                    mintable: mintable,
                    original_symbol: symbol,
                    owner: this.address,
                    symbol: symbol,
                    total_supply: total_supply,
                    whole_name: whole_name
                  }
                }];
                _context18.next = 7;
                return this.buildTransaction(msg, msg, memo, defaultFee, sequenceNumber);

              case 7:
                signedTx = _context18.sent;
                _context18.next = 10;
                return this.sendTransaction(signedTx);

              case 10:
                res = _context18.sent;
                return _context18.abrupt("return", res);

              case 12:
              case "end":
                return _context18.stop();
            }
          }
        }, _callee18, this);
      }));

      function sendTokenIssueTransaction(_x22, _x23, _x24) {
        return _sendTokenIssueTransaction.apply(this, arguments);
      }

      return sendTokenIssueTransaction;
    }()
    /**
     * Send TokenBurnTransaction.
     * @param {String} token
     * @param {String} amount
     * @param {String} memo
     * @param {Number} sequenceNumber
     * @return {Object} response
     */

  }, {
    key: "sendTokenBurnTransaction",
    value: function () {
      var _sendTokenBurnTransaction = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee19(token, amount) {
        var memo,
            sequenceNumber,
            msg,
            signedTx,
            res,
            _args19 = arguments;
        return regeneratorRuntime.wrap(function _callee19$(_context19) {
          while (1) {
            switch (_context19.prev = _context19.next) {
              case 0:
                memo = _args19.length > 2 && _args19[2] !== undefined ? _args19[2] : "";
                sequenceNumber = _args19.length > 3 && _args19[3] !== undefined ? _args19[3] : null;
                msg = [{
                  type: "okexchain/token/MsgBurn",
                  value: {
                    amount: {
                      amount: this.formatNumber(amount),
                      denom: token
                    },
                    owner: this.address
                  }
                }];
                _context19.next = 5;
                return this.buildTransaction(msg, msg, memo, defaultFee, sequenceNumber);

              case 5:
                signedTx = _context19.sent;
                _context19.next = 8;
                return this.sendTransaction(signedTx);

              case 8:
                res = _context19.sent;
                return _context19.abrupt("return", res);

              case 10:
              case "end":
                return _context19.stop();
            }
          }
        }, _callee19, this);
      }));

      function sendTokenBurnTransaction(_x25, _x26) {
        return _sendTokenBurnTransaction.apply(this, arguments);
      }

      return sendTokenBurnTransaction;
    }()
    /**
     * Send TokenMintTransaction.
     * @param {String} token
     * @param {String} amount
     * @param {String} memo
     * @param {Number} sequenceNumber
     * @return {Object} response
     */

  }, {
    key: "sendTokenMintTransaction",
    value: function () {
      var _sendTokenMintTransaction = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee20(token, amount) {
        var memo,
            sequenceNumber,
            msg,
            signedTx,
            res,
            _args20 = arguments;
        return regeneratorRuntime.wrap(function _callee20$(_context20) {
          while (1) {
            switch (_context20.prev = _context20.next) {
              case 0:
                memo = _args20.length > 2 && _args20[2] !== undefined ? _args20[2] : "";
                sequenceNumber = _args20.length > 3 && _args20[3] !== undefined ? _args20[3] : null;
                msg = [{
                  type: "okexchain/token/MsgMint",
                  value: {
                    amount: {
                      amount: this.formatNumber(amount),
                      denom: token
                    },
                    owner: this.address
                  }
                }];
                _context20.next = 5;
                return this.buildTransaction(msg, msg, memo, defaultFee, sequenceNumber);

              case 5:
                signedTx = _context20.sent;
                _context20.next = 8;
                return this.sendTransaction(signedTx);

              case 8:
                res = _context20.sent;
                return _context20.abrupt("return", res);

              case 10:
              case "end":
                return _context20.stop();
            }
          }
        }, _callee20, this);
      }));

      function sendTokenMintTransaction(_x27, _x28) {
        return _sendTokenMintTransaction.apply(this, arguments);
      }

      return sendTokenMintTransaction;
    }()
    /**
     * Send RegisterDexOperatorTransaction.
     * @param {String} website
     * @param {String} handling_fee_address
     * @param {String} memo
     * @param {Number} sequenceNumber
     * @return {Object} response
     */

  }, {
    key: "sendRegisterDexOperatorTransaction",
    value: function () {
      var _sendRegisterDexOperatorTransaction = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee21(website, handling_fee_address) {
        var memo,
            sequenceNumber,
            msg,
            signedTx,
            res,
            _args21 = arguments;
        return regeneratorRuntime.wrap(function _callee21$(_context21) {
          while (1) {
            switch (_context21.prev = _context21.next) {
              case 0:
                memo = _args21.length > 2 && _args21[2] !== undefined ? _args21[2] : "";
                sequenceNumber = _args21.length > 3 && _args21[3] !== undefined ? _args21[3] : null;

                if (crypto.validateAddress(handling_fee_address)) {
                  _context21.next = 4;
                  break;
                }

                throw new Error("invalid handling_fee_address");

              case 4:
                msg = [{
                  type: "okexchain/dex/CreateOperator",
                  value: {
                    handling_fee_address: handling_fee_address,
                    owner: this.address,
                    website: website
                  }
                }];
                _context21.next = 7;
                return this.buildTransaction(msg, msg, memo, defaultFee, sequenceNumber);

              case 7:
                signedTx = _context21.sent;
                _context21.next = 10;
                return this.sendTransaction(signedTx);

              case 10:
                res = _context21.sent;
                return _context21.abrupt("return", res);

              case 12:
              case "end":
                return _context21.stop();
            }
          }
        }, _callee21, this);
      }));

      function sendRegisterDexOperatorTransaction(_x29, _x30) {
        return _sendRegisterDexOperatorTransaction.apply(this, arguments);
      }

      return sendRegisterDexOperatorTransaction;
    }()
    /**
     * Send ListTokenPairTransaction.
     * @param {String} base_asset
     * @param {String} quote_asset
     * @param {String} init_price
     * @param {String} memo
     * @param {Number} sequenceNumber
     * @return {Object} response
     */

  }, {
    key: "sendListTokenPairTransaction",
    value: function () {
      var _sendListTokenPairTransaction = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee22(base_asset, quote_asset, init_price) {
        var memo,
            sequenceNumber,
            msg,
            signedTx,
            res,
            _args22 = arguments;
        return regeneratorRuntime.wrap(function _callee22$(_context22) {
          while (1) {
            switch (_context22.prev = _context22.next) {
              case 0:
                memo = _args22.length > 3 && _args22[3] !== undefined ? _args22[3] : "";
                sequenceNumber = _args22.length > 4 && _args22[4] !== undefined ? _args22[4] : null;
                msg = [{
                  type: "okexchain/dex/MsgList",
                  value: {
                    init_price: this.formatNumber(init_price),
                    list_asset: base_asset,
                    owner: this.address,
                    quote_asset: quote_asset
                  }
                }];
                _context22.next = 5;
                return this.buildTransaction(msg, msg, memo, defaultFee, sequenceNumber);

              case 5:
                signedTx = _context22.sent;
                _context22.next = 8;
                return this.sendTransaction(signedTx);

              case 8:
                res = _context22.sent;
                return _context22.abrupt("return", res);

              case 10:
              case "end":
                return _context22.stop();
            }
          }
        }, _callee22, this);
      }));

      function sendListTokenPairTransaction(_x31, _x32, _x33) {
        return _sendListTokenPairTransaction.apply(this, arguments);
      }

      return sendListTokenPairTransaction;
    }()
    /**
     * Send AddProductDepositTransaction.
     * @param {String} amount
     * @param {String} product
     * @param {String} memo
     * @param {Number} sequenceNumber
     * @return {Object} response
     */

  }, {
    key: "sendAddProductDepositTransaction",
    value: function () {
      var _sendAddProductDepositTransaction = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee23(amount, product) {
        var memo,
            sequenceNumber,
            coin,
            msg,
            signedTx,
            res,
            _args23 = arguments;
        return regeneratorRuntime.wrap(function _callee23$(_context23) {
          while (1) {
            switch (_context23.prev = _context23.next) {
              case 0:
                memo = _args23.length > 2 && _args23[2] !== undefined ? _args23[2] : "";
                sequenceNumber = _args23.length > 3 && _args23[3] !== undefined ? _args23[3] : null;
                coin = {
                  amount: this.formatNumber(amount),
                  denom: nativeDenom
                };
                msg = [{
                  type: "okexchain/dex/MsgDeposit",
                  value: {
                    amount: coin,
                    depositor: this.address,
                    product: product
                  }
                }];
                _context23.next = 6;
                return this.buildTransaction(msg, msg, memo, defaultFee, sequenceNumber);

              case 6:
                signedTx = _context23.sent;
                _context23.next = 9;
                return this.sendTransaction(signedTx);

              case 9:
                res = _context23.sent;
                return _context23.abrupt("return", res);

              case 11:
              case "end":
                return _context23.stop();
            }
          }
        }, _callee23, this);
      }));

      function sendAddProductDepositTransaction(_x34, _x35) {
        return _sendAddProductDepositTransaction.apply(this, arguments);
      }

      return sendAddProductDepositTransaction;
    }()
    /**
     * Send WithdrawProductDepositTransaction.
     * @param {String} amount
     * @param {String} product
     * @param {String} memo
     * @param {Number} sequenceNumber
     * @return {Object} response
     */

  }, {
    key: "sendWithdrawProductDepositTransaction",
    value: function () {
      var _sendWithdrawProductDepositTransaction = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee24(amount, product) {
        var memo,
            sequenceNumber,
            coin,
            msg,
            signedTx,
            res,
            _args24 = arguments;
        return regeneratorRuntime.wrap(function _callee24$(_context24) {
          while (1) {
            switch (_context24.prev = _context24.next) {
              case 0:
                memo = _args24.length > 2 && _args24[2] !== undefined ? _args24[2] : "";
                sequenceNumber = _args24.length > 3 && _args24[3] !== undefined ? _args24[3] : null;
                coin = {
                  amount: this.formatNumber(amount),
                  denom: nativeDenom
                };
                msg = [{
                  type: "okexchain/dex/MsgWithdraw",
                  value: {
                    amount: coin,
                    depositor: this.address,
                    product: product
                  }
                }];
                _context24.next = 6;
                return this.buildTransaction(msg, msg, memo, defaultFee, sequenceNumber);

              case 6:
                signedTx = _context24.sent;
                _context24.next = 9;
                return this.sendTransaction(signedTx);

              case 9:
                res = _context24.sent;
                return _context24.abrupt("return", res);

              case 11:
              case "end":
                return _context24.stop();
            }
          }
        }, _callee24, this);
      }));

      function sendWithdrawProductDepositTransaction(_x36, _x37) {
        return _sendWithdrawProductDepositTransaction.apply(this, arguments);
      }

      return sendWithdrawProductDepositTransaction;
    }()
    /**
     * Send AddLiquidityTransaction.
     * @param {Number} min_liquidity
     * @param {Number} max_base_amount
     * @param {String} base_token
     * @param {Number} quote_amount
     * @param {String} quote_token
     * @param {String} deadline
     * @param {String} memo
     * @param {Number} sequenceNumber
     * @return {Object} response
     */

  }, {
    key: "sendAddLiquidityTransaction",
    value: function () {
      var _sendAddLiquidityTransaction = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee25(min_liquidity, max_base_amount, base_token, quote_amount, quote_token, deadline) {
        var memo,
            sequenceNumber,
            base_coin,
            quote_coin,
            msg,
            signedTx,
            res,
            _args25 = arguments;
        return regeneratorRuntime.wrap(function _callee25$(_context25) {
          while (1) {
            switch (_context25.prev = _context25.next) {
              case 0:
                memo = _args25.length > 6 && _args25[6] !== undefined ? _args25[6] : '';
                sequenceNumber = _args25.length > 7 && _args25[7] !== undefined ? _args25[7] : null;
                base_coin = {
                  amount: this.formatNumber(max_base_amount),
                  denom: base_token
                };
                quote_coin = {
                  amount: this.formatNumber(quote_amount),
                  denom: quote_token
                };
                msg = [{
                  type: "okexchain/ammswap/MsgAddLiquidity",
                  value: {
                    deadline: deadline,
                    max_base_amount: base_coin,
                    min_liquidity: this.formatNumber(min_liquidity),
                    quote_amount: quote_coin,
                    sender: this.address
                  }
                }];
                _context25.next = 7;
                return this.buildTransaction(msg, msg, memo, defaultFee, sequenceNumber);

              case 7:
                signedTx = _context25.sent;
                _context25.next = 10;
                return this.sendTransaction(signedTx);

              case 10:
                res = _context25.sent;
                return _context25.abrupt("return", res);

              case 12:
              case "end":
                return _context25.stop();
            }
          }
        }, _callee25, this);
      }));

      function sendAddLiquidityTransaction(_x38, _x39, _x40, _x41, _x42, _x43) {
        return _sendAddLiquidityTransaction.apply(this, arguments);
      }

      return sendAddLiquidityTransaction;
    }()
    /**
     * Send RemoveLiquidityTransaction.
     * @param {Number} liquidity
     * @param {Number} min_base_amount
     * @param {String} base_token
     * @param {Number} min_quote_amount
     * @param {String} quote_token
     * @param {String} deadline
     * @param {String} memo
     * @param {Number} sequenceNumber
     * @return {Object} response
     */

  }, {
    key: "sendRemoveLiquidityTransaction",
    value: function () {
      var _sendRemoveLiquidityTransaction = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee26(liquidity, min_base_amount, base_token, min_quote_amount, quote_token, deadline) {
        var memo,
            sequenceNumber,
            base_coin,
            quote_coin,
            msg,
            signedTx,
            res,
            _args26 = arguments;
        return regeneratorRuntime.wrap(function _callee26$(_context26) {
          while (1) {
            switch (_context26.prev = _context26.next) {
              case 0:
                memo = _args26.length > 6 && _args26[6] !== undefined ? _args26[6] : '';
                sequenceNumber = _args26.length > 7 && _args26[7] !== undefined ? _args26[7] : null;
                base_coin = {
                  amount: this.formatNumber(min_base_amount),
                  denom: base_token
                };
                quote_coin = {
                  amount: this.formatNumber(min_quote_amount),
                  denom: quote_token
                };
                msg = [{
                  type: "okexchain/ammswap/MsgRemoveLiquidity",
                  value: {
                    deadline: deadline,
                    liquidity: this.formatNumber(liquidity),
                    min_base_amount: base_coin,
                    min_quote_amount: quote_coin,
                    sender: this.address
                  }
                }];
                _context26.next = 7;
                return this.buildTransaction(msg, msg, memo, defaultFee, sequenceNumber);

              case 7:
                signedTx = _context26.sent;
                _context26.next = 10;
                return this.sendTransaction(signedTx);

              case 10:
                res = _context26.sent;
                return _context26.abrupt("return", res);

              case 12:
              case "end":
                return _context26.stop();
            }
          }
        }, _callee26, this);
      }));

      function sendRemoveLiquidityTransaction(_x44, _x45, _x46, _x47, _x48, _x49) {
        return _sendRemoveLiquidityTransaction.apply(this, arguments);
      }

      return sendRemoveLiquidityTransaction;
    }()
    /**
     * Send CreateExchangeTransaction.
     * @param {String} Token0Name
     * @param {String} Token1Name
     * @param {String} memo
     * @param {Number} sequenceNumber
     * @return {Object} response
     */

  }, {
    key: "sendCreateExchangeTransaction",
    value: function () {
      var _sendCreateExchangeTransaction = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee27(Token0Name, Token1Name) {
        var memo,
            sequenceNumber,
            msg,
            signedTx,
            res,
            _args27 = arguments;
        return regeneratorRuntime.wrap(function _callee27$(_context27) {
          while (1) {
            switch (_context27.prev = _context27.next) {
              case 0:
                memo = _args27.length > 2 && _args27[2] !== undefined ? _args27[2] : '';
                sequenceNumber = _args27.length > 3 && _args27[3] !== undefined ? _args27[3] : null;
                msg = [{
                  type: "okexchain/ammswap/MsgCreateExchange",
                  value: {
                    sender: this.address,
                    token0_name: Token0Name,
                    token1_name: Token1Name
                  }
                }];
                _context27.next = 5;
                return this.buildTransaction(msg, msg, memo, defaultFee, sequenceNumber);

              case 5:
                signedTx = _context27.sent;
                _context27.next = 8;
                return this.sendTransaction(signedTx);

              case 8:
                res = _context27.sent;
                return _context27.abrupt("return", res);

              case 10:
              case "end":
                return _context27.stop();
            }
          }
        }, _callee27, this);
      }));

      function sendCreateExchangeTransaction(_x50, _x51) {
        return _sendCreateExchangeTransaction.apply(this, arguments);
      }

      return sendCreateExchangeTransaction;
    }()
    /**
     * Send SwapTokenTransaction.
     * @param {Number} sold_token_amount
     * @param {String} sold_token
     * @param {Number} min_bought_token_amount
     * @param {String} bought_token
     * @param {String} deadline
     * @param {String} recipient
     * @param {String} memo
     * @param {Number} sequenceNumber
     * @return {Object} response
     */

  }, {
    key: "sendSwapTokenTransaction",
    value: function () {
      var _sendSwapTokenTransaction = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee28(sold_token_amount, sold_token, min_bought_token_amount, bought_token, deadline, recipient) {
        var memo,
            sequenceNumber,
            sold_coin,
            bought_coin,
            msg,
            signedTx,
            res,
            _args28 = arguments;
        return regeneratorRuntime.wrap(function _callee28$(_context28) {
          while (1) {
            switch (_context28.prev = _context28.next) {
              case 0:
                memo = _args28.length > 6 && _args28[6] !== undefined ? _args28[6] : '';
                sequenceNumber = _args28.length > 7 && _args28[7] !== undefined ? _args28[7] : null;
                sold_coin = {
                  amount: this.formatNumber(sold_token_amount),
                  denom: sold_token
                };
                bought_coin = {
                  amount: this.formatNumber(min_bought_token_amount),
                  denom: bought_token
                };
                msg = [{
                  type: "okexchain/ammswap/MsgSwapToken",
                  value: {
                    deadline: deadline,
                    min_bought_token_amount: bought_coin,
                    recipient: recipient,
                    sender: this.address,
                    sold_token_amount: sold_coin
                  }
                }];
                _context28.next = 7;
                return this.buildTransaction(msg, msg, memo, defaultFee, sequenceNumber);

              case 7:
                signedTx = _context28.sent;
                _context28.next = 10;
                return this.sendTransaction(signedTx);

              case 10:
                res = _context28.sent;
                return _context28.abrupt("return", res);

              case 12:
              case "end":
                return _context28.stop();
            }
          }
        }, _callee28, this);
      }));

      function sendSwapTokenTransaction(_x52, _x53, _x54, _x55, _x56, _x57) {
        return _sendSwapTokenTransaction.apply(this, arguments);
      }

      return sendSwapTokenTransaction;
    }()
  }]);

  return OKEXChainClient;
}();

exports.OKEXChainClient = OKEXChainClient;
